feat(transpile) - auto-populating the .has tree

ttodua opened the pull request:
without .features PR this PR would cause to loose some details, like `createStopOrder = true` and etc simulated has props.

ttodua committed:
feat(transpile) - auto-populating the .has tree

ttodua committed:
exchange ts

ttodua committed:
ex revert

ttodua committed:
p2b example

ttodua committed:
rev has

ttodua committed:
var rev

ttodua committed:
transpiler update

ttodua committed:
undef

ttodua committed:
p2b rev

ttodua committed:
comment

ttodua committed:
emulated support

ttodua committed:
mark/index/premium

ttodua committed:
sandbox

ttodua committed:
PremiumIndex

ttodua committed:
fill `has` autogeneration with remaining methods

ttodua committed:
add all order-related methods

carlosmiei commented on build/transpile.js:
```diff
@@ -119,6 +119,9 @@ class Transpiler {
             [ /\.parseTimeInForce /g, '.parse_time_in_force'],
             [ /\.parseTradingFees /g, '.parse_trading_fees'],
             [ /\.describeData /g, '.describe_data'],
+            [ /\.defaultHas /g, '.default_has'],
+            [ /\.getOwnMethods /g, '.get_own_methods'],
+            [ /\.populateHasTree /g, '.populate_has_tree'],
             [ /\.randNumber /g, '.rand_number'],
```
why is this needed?

ttodua replied:
ah, this is leftover from previous approach, removed it atm

carlosmiei commented on build/transpile.js:
```diff
@@ -889,6 +892,9 @@ class Transpiler {
 
     sortExchangeCapabilities (code) {
+        const baseExchange = this.getBaseClass ()
+        const defaultDescribe = baseExchange.describe ();
```
hum, this assumes the code was transpiled first to JS? Usually that requirement was not needed

ttodua replied:
no, this method was in place before transpilation happens (sorts the `has` dict) and after that starts transpilation.
if you look this line where this method is called: 
https://github.com/ccxt/ccxt/blob/f22d9b2f87a9739cd156676a08fa4090ea2f0809/build/transpile.js#L1284

down below 2 lines then the .ts file is overwritten, and transpilation starts then with the "sorted" code.

ttodua replied:
I avoided to include all changes that happen after running transpiler, so if you want to see the changes it does to any exchange, you might run the transpiler and see diff for `binance` or any exchange

ttodua replied:
lmk if we should close this PR

ttodua committed:
Update transpile.js

ttodua committed:
Merge branch 'ccxt:master' into has-autogeneration

ttodua committed:
master

ttodua committed:
Merge branch 'master' of github.com:ccxt/ccxt into has-autogeneration

ttodua committed:
tranpsile ts upd

ttodua committed:
Merge branch 'has-autogeneration' of github.com:ttodua/ccxt into has-autogeneration

ttodua committed:
Update transpile.ts

ttodua committed:
Merge branch 'master' into has-autogeneration

ttodua commented:
@carlosmiei

ttodua committed:
-
Merge branch 'master' of github.com:ccxt/ccxt into has-autogeneration

# Conflicts:
#	build/transpile.ts
#	ts/src/p2b.ts

ttodua committed:
transpiler upd

ttodua committed:
Merge branch 'has-autogeneration' of github.com:ttodua/ccxt into has-autogeneration

ttodua committed:
-
Revert "Merge branch 'has-autogeneration' of github.com:ttodua/ccxt into has-autogeneration"

This reverts commit c7c418208f9d9936d82d566d47402df37901f2f9, reversing
changes made to 5c3928ee55a99da56d252296261179668cd8231c.

ttodua commented:
converting to draft bcz of inactivity

ttodua committed:
Merge branch 'master' of github.com:ccxt/ccxt into has-autogeneration

ttodua committed:
update

ttodua committed:
Merge branch 'master' of github.com:ccxt/ccxt into has-autogeneration

ttodua committed:
some updates

ttodua committed:
minor

ttodua committed:
რევ

ttodua commented on build/transpile.ts:
```diff
@@ -976,6 +977,69 @@ class Transpiler {
+            'createTrailingAmountOrderWs',
+            'createTrailingPercentOrder',
+            'createTrailingPercentOrderWs',
+            'createMarketOrderWithCost',
+            'createMarketOrderWithCostWs',
+            'createLimitOrder',
+            'createMarketOrder',
+            'createLimitBuyOrder',
```
done
